File: apps/visualizer/backend/jobs/runner.py
import threading

from database import (
    add_job_log,
    get_job,
    job_log_has_message,
    make_connection_for_path,
    update_job_field,
    update_job_status,
)

from .checkpoint import merge_checkpoint_into_metadata


class JobRunner:
    """Coordinate job lifecycle across the processor thread and workers.

    Each thread that mutates the jobs DB through the runner should use its
    *own* sqlite3 connection. The runner carries a ``db_path`` for that
    purpose; call :meth:`thread_db` (or pass one of the logging helpers the
    runner exposes) and you'll get a thread-local connection. ``self.db``
    remains the connection belonging to whichever thread constructed the
    runner (typically the processor thread) so existing single-threaded
    call sites are unaffected.
    """

    def __init__(self, db, emit_progress=None, *, db_path: str | None = None):
        self.db = db
        # Path is optional for legacy tests that build a JobRunner directly
        # from a connection. Production paths pass ``db_path`` so the runner
        # can hand out thread-local connections to worker threads.
        self.db_path = db_path
        self.emit_progress = emit_progress or (lambda *args: None)
        self.active_jobs: dict[str, threading.Event] = {}

    # ------------------------------------------------------------------
    # Thread-local DB handle
    # ------------------------------------------------------------------
    def thread_db(self):
        """Return a sqlite3 connection owned by the calling thread.

        Worker threads must never share ``self.db`` — that was the root
        cause of the 3h main-thread stall observed on job 50710bf6, where
        four worker log callbacks and the coordinator loop all serialized
        on a single connection's internal mutex.
        """
        if self.db_path is None:
            # Legacy callers that built the runner without a path fall back
            # to the shared connection. Only safe from the owning thread,
            # and that is exactly how the old tests use it.
            return self.db
        return make_connection_for_path(self.db_path)

    def log_from_worker(self, job_id: str, level: str, message: str) -> None:
        """Append a log entry using the *calling thread's* connection.

        Workers pump hundreds of debug messages per minute; routing them
        through ``self.db`` would re-serialize everything on one mutex.
        """
        add_job_log(self.thread_db(), job_id, level, message)

    def start_job(self, job_id: str, job_type: str, metadata: dict) -> bool:
        """Mark job as running. Returns False if the job row is gone or already cancelled."""
        row = get_job(self.db, job_id)
        if not row or row.get('status') == 'cancelled':
            return False
        cancel_event = threading.Event()
        self.active_jobs[job_id] = cancel_event
        update_job_status(self.db, job_id, 'running', progress=0, current_step='Starting...')
        add_job_log(self.db, job_id, 'info', f'Job {job_type} started')
        return True

    def update_progress(self, job_id: str, progress: int, current_step: str):
        """Update job progress. No-op if the job has been cancelled or already completed."""
        if self.is_cancelled(job_id):
            return
        row = get_job(self.db, job_id)
        if row and row.get('status') in ('completed', 'cancelled'):
            return
        update_job_status(self.db, job_id, 'running', progress=progress, current_step=current_step)
        add_job_log(self.db, job_id, 'info', current_step)
        self.emit_progress(job_id, progress, current_step)

    def complete_job(self, job_id: str, result: dict):
        """Mark job as completed."""
        row = get_job(self.db, job_id)
        if row and row.get('status') == 'cancelled':
            self.clear_cancel_registration(job_id)
            return
        update_job_status(self.db, job_id, 'completed', progress=100)
        add_job_log(self.db, job_id, 'info', 'Job completed successfully')

        update_job_field(self.db, job_id, 'result', result)
        self.db.execute(
            "UPDATE jobs SET error_severity = NULL WHERE id = ?",
            (job_id,),
        )
        self.db.commit()
        self.clear_cancel_registration(job_id)

    def fail_job(self, job_id: str, error: str, *, severity: str = 'error') -> None:
        """Mark job as failed."""
        if severity not in ('warning', 'error', 'critical'):
            severity = 'error'
        row = get_job(self.db, job_id)
        if row and row.get('status') == 'cancelled':
            self.clear_cancel_registration(job_id)
            return
        update_job_status(self.db, job_id, 'failed')
        add_job_log(self.db, job_id, 'error', error)

        self.db.execute(
            "UPDATE jobs SET error = ?, error_severity = ? WHERE id = ?",
            (error, severity, job_id),
        )
        self.db.commit()
        self.clear_cancel_registration(job_id)

    def is_cancelled(self, job_id: str) -> bool:
        ev = self.active_jobs.get(job_id)
        if ev and ev.is_set():
            return True
        row = get_job(self.db, job_id)
        if row and row.get('status') == 'cancelled':
            if ev:
                ev.set()
            return True
        return False

    def clear_cancel_registration(self, job_id: str) -> None:
        self.active_jobs.pop(job_id, None)

    def persist_checkpoint(self, job_id: str, checkpoint_body: dict) -> None:
        """Merge versioned checkpoint data into ``jobs.metadata`` (jobs DB)."""
        row = get_job(self.db, job_id)
        if not row:
            return
        meta = row.get("metadata") or {}
        if not isinstance(meta, dict):
            meta = {}
        new_meta = merge_checkpoint_into_metadata(meta, checkpoint_body)
        update_job_field(self.db, job_id, "metadata", new_meta)

    def clear_checkpoint(self, job_id: str) -> None:
        """Remove resume checkpoint from job metadata (e.g. after successful completion)."""
        row = get_job(self.db, job_id)
        if not row:
            return
        meta = row.get("metadata") or {}
        if not isinstance(meta, dict):
            meta = {}
        new_meta = {**meta, "checkpoint": None}
        update_job_field(self.db, job_id, "metadata", new_meta)

    def signal_cancel(self, job_id: str) -> None:
        """Set the cooperative cancel flag for a running job (no DB writes)."""
        if job_id in self.active_jobs:
            self.active_jobs[job_id].set()

    def finalize_cancelled(self, job_id: str) -> None:
        """Clear in-memory cancel state and ensure DB reflects cooperative cancel."""
        self.clear_cancel_registration(job_id)
        # Avoid pulling the full logs — ``job_log_has_message`` is an
        # index-backed ``LIMIT 1`` lookup so this is fast even when a job
        # has accumulated 20k+ log rows.
        row = get_job(self.db, job_id, logs_limit=0)
        if not row:
            return
        stop_msg = 'Job stopped after cancel request'
        has_stop_msg = job_log_has_message(self.db, job_id, stop_msg)
        status = row.get('status')
        if status == 'running':
            update_job_status(self.db, job_id, 'cancelled')
            if not has_stop_msg:
                add_job_log(self.db, job_id, 'info', stop_msg)
        elif status == 'cancelled' and not has_stop_msg:
            add_job_log(self.db, job_id, 'info', stop_msg)

File: apps/visualizer/backend/src/jobs/runner.ts
/**
 * Coordinate the job lifecycle between the processor loop and the handlers.
 *
 * One thread, one SQLite connection; handlers that need real parallelism use
 * `worker_threads` with their own connection. WAL plus `busy_timeout` is the
 * coordination point across processes.
 */
import type { Db } from '../db/connection.js';
import {
  addJobLog,
  getJob,
  jobLogHasMessage,
  updateJobField,
  updateJobStatus,
  type JobLogLevel,
} from '../db/jobs/jobs.js';
import { CHECKPOINT_VERSION } from './checkpoint.js';

export type EmitProgress = (jobId: string, progress: number, currentStep: string) => void;

export type ErrorSeverity = 'warning' | 'error' | 'critical';

const STOP_MESSAGE = 'Job stopped after cancel request';

export class JobRunner {
  private readonly db: Db;
  private readonly emitProgress: EmitProgress;

  /**
   * Cooperative cancel flags for in-flight jobs.
   *
   * A boolean box rather than an event: handlers poll `isCancelled` between
   * units of work, so there is nothing to wait on.
   */
  private readonly activeJobs = new Map<string, { cancelled: boolean }>();

  /**
   * Last `(progress, step)` logged per job.
   *
   * Progress callbacks fire far more often than the step changes, and without
   * this a batch job writes thousands of identical log lines.
   */
  private readonly lastProgressLog = new Map<string, string>();

  constructor(db: Db, emitProgress: EmitProgress = () => {}) {
    this.db = db;
    this.emitProgress = emitProgress;
  }

  /** Append a log entry. Kept as a method so handlers need no DB handle. */
  log(jobId: string, level: JobLogLevel, message: string): void {
    addJobLog(this.db, jobId, level, message);
  }

  /** The job's metadata, including any `checkpoint` a previous run persisted. */
  readMetadata(jobId: string): Record<string, unknown> {
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    const meta = row?.metadata;
    return meta && typeof meta === 'object' && !Array.isArray(meta)
      ? (meta as Record<string, unknown>)
      : {};
  }

  /** Whether the job has already settled, e.g. failed from inside a handler. */
  hasFailed(jobId: string): boolean {
    return getJob(this.db, jobId, { logsLimit: 0 })?.status === 'failed';
  }

  /**
   * Mark a job running. Returns false when the row is gone or already cancelled,
   * which is the check that stops a job cancelled while queued from starting.
   */
  startJob(jobId: string, jobType: string): boolean {
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (!row || row.status === 'cancelled') return false;
    this.activeJobs.set(jobId, { cancelled: false });
    this.lastProgressLog.delete(jobId);
    updateJobStatus(this.db, jobId, 'running', { progress: 0, currentStep: 'Starting...' });
    addJobLog(this.db, jobId, 'info', `Job ${jobType} started`);
    return true;
  }

  /** Update progress. A no-op once the job is cancelled or already finished. */
  updateProgress(jobId: string, progress: number, currentStep: string): void {
    if (this.isCancelled(jobId)) return;
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (row && (row.status === 'completed' || row.status === 'cancelled')) return;

    updateJobStatus(this.db, jobId, 'running', { progress, currentStep });
    const logKey = `${progress}\u0000${currentStep}`;
    if (this.lastProgressLog.get(jobId) !== logKey) {
      addJobLog(this.db, jobId, 'info', currentStep);
      this.lastProgressLog.set(jobId, logKey);
    }
    this.emitProgress(jobId, progress, currentStep);
  }

  /**
   * Name the stage a composite job is in, without touching its progress.
   *
   * `updateProgress` also writes `current_step`, but it is called once per unit of
   * work; this is called once per stage, so the label survives being overwritten
   * by the per-item messages only in the sense that the UI reads both.
   */
  setCurrentStep(jobId: string, currentStep: string): void {
    updateJobField(this.db, jobId, 'current_step', currentStep);
  }

  /**
   * Mark a job completed and store its result.
   *
   * A job cancelled while the handler was finishing stays cancelled: the user's
   * decision wins over a late success.
   */
  completeJob(jobId: string, result: unknown): void {
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (row && row.status === 'cancelled') {
      this.clearCancelRegistration(jobId);
      return;
    }
    updateJobStatus(this.db, jobId, 'completed', { progress: 100 });
    addJobLog(this.db, jobId, 'info', 'Job completed successfully');
    updateJobField(this.db, jobId, 'result', result);
    this.db.prepare('UPDATE jobs SET error_severity = NULL WHERE id = ?').run(jobId);
    this.lastProgressLog.delete(jobId);
    this.clearCancelRegistration(jobId);
  }

  /** Mark a job failed, with a severity the UI colours by. */
  failJob(jobId: string, error: string, severity: ErrorSeverity = 'error'): void {
    const resolved: ErrorSeverity =
      severity === 'warning' || severity === 'error' || severity === 'critical'
        ? severity
        : 'error';
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (row && row.status === 'cancelled') {
      this.clearCancelRegistration(jobId);
      return;
    }
    updateJobStatus(this.db, jobId, 'failed');
    addJobLog(this.db, jobId, 'error', error);
    this.db
      .prepare('UPDATE jobs SET error = ?, error_severity = ? WHERE id = ?')
      .run(error, resolved, jobId);
    this.lastProgressLog.delete(jobId);
    this.clearCancelRegistration(jobId);
  }

  /**
   * Whether the job has been cancelled.
   *
   * Checks the in-memory flag first, then the database — the flag is set by
   * `signalCancel` in the same process, but a cancel can also arrive as a plain
   * status write, and a long-running handler has to notice either.
   */
  isCancelled(jobId: string): boolean {
    const flag = this.activeJobs.get(jobId);
    if (flag?.cancelled) return true;
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (row && row.status === 'cancelled') {
      if (flag) flag.cancelled = true;
      return true;
    }
    return false;
  }

  clearCancelRegistration(jobId: string): void {
    this.activeJobs.delete(jobId);
    this.lastProgressLog.delete(jobId);
  }

  /** Set the cooperative cancel flag. Deliberately writes nothing to the DB. */
  signalCancel(jobId: string): void {
    const flag = this.activeJobs.get(jobId);
    if (flag) flag.cancelled = true;
  }

  /**
   * Merge versioned checkpoint data into `jobs.metadata`.
   *
   * The version is stamped here rather than by the caller, because
   * `recoverOrphanedJobs` re-queues on it — a handler that forgot it would have
   * its job failed on restart instead of resumed.
   */
  persistCheckpoint(jobId: string, checkpointBody: Record<string, unknown>): void {
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (!row) return;
    const meta = row.metadata && typeof row.metadata === 'object' ? row.metadata : {};
    updateJobField(this.db, jobId, 'metadata', {
      ...meta,
      checkpoint: { checkpoint_version: CHECKPOINT_VERSION, ...checkpointBody },
    });
  }

  /** Drop the resume checkpoint, e.g. after a clean completion. */
  clearCheckpoint(jobId: string): void {
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (!row) return;
    const meta = row.metadata && typeof row.metadata === 'object' ? row.metadata : {};
    updateJobField(this.db, jobId, 'metadata', { ...meta, checkpoint: null });
  }

  /**
   * Settle a cooperatively-cancelled job.
   *
   * `logsLimit: 0` and the indexed marker lookup keep this cheap even for a job
   * that accumulated 20,000 log rows — loading the history just to check for one
   * message is what this avoids.
   */
  finalizeCancelled(jobId: string): void {
    this.clearCancelRegistration(jobId);
    const row = getJob(this.db, jobId, { logsLimit: 0 });
    if (!row) return;

    const hasStopMessage = jobLogHasMessage(this.db, jobId, STOP_MESSAGE);
    if (row.status === 'running') {
      updateJobStatus(this.db, jobId, 'cancelled');
      if (!hasStopMessage) addJobLog(this.db, jobId, 'info', STOP_MESSAGE);
    } else if (row.status === 'cancelled' && !hasStopMessage) {
      addJobLog(this.db, jobId, 'info', STOP_MESSAGE);
    }
  }
}
